crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -1123,47 +1125,51 @@ impl OlmMachine {
     /// * `room_id` - The ID of the room where the event was sent to.
     pub async fn decrypt_room_event(
         &self,
-        event: &OriginalSyncRoomEncryptedEvent,
+        event: &Raw<EncryptedEvent>,
         room_id: &RoomId,
     ) -> MegolmResult<RoomEvent> {
-        match &event.content.scheme {
-            EncryptedEventScheme::MegolmV1AesSha2(c) => {
-                match self.decrypt_megolm_v1_event(room_id, event, c).await {
-                    Ok(r) => Ok(r),
-                    Err(e) => {
-                        #[allow(deprecated)]
-                        if let MegolmError::MissingRoomKey = e {
-                            // TODO log the withheld reason if we have one.
-                            debug!(
-                                sender = event.sender.as_str(),
-                                room_id = room_id.as_str(),
-                                sender_key = c.sender_key.as_str(),
-                                session_id = c.session_id.as_str(),
-                                "Failed to decrypt a room event, the room key is missing"
-                            );
-                        } else {
-                            warn!(
-                                sender = event.sender.as_str(),
-                                room_id = room_id.as_str(),
-                                sender_key = c.sender_key.as_str(),
-                                session_id = c.session_id.as_str(),
-                                error = ?e,
-                                "Failed to decrypt a room event"
-                            );
-                        }
+        let event = event.deserialize()?;
 
-                        Err(e)
-                    }
-                }
-            }
-            algorithm => {
+        let content = match &event.content.scheme {
+            RoomEventEncryptionScheme::MegolmV1AesSha2(c) => c.into(),
+            RoomEventEncryptionScheme::Unknown(c) => {
                 warn!(
                     sender = event.sender.as_str(),
                     room_id = room_id.as_str(),
-                    ?algorithm,
+                    algorithm =? c.algorithm,
                     "Received an encrypted room event with an unsupported algorithm"
                 );
-                Err(EventError::UnsupportedAlgorithm.into())
+
+                return Err(EventError::UnsupportedAlgorithm.into());
+            }
+        };
+
+        match self.decrypt_megolm_events(room_id, &event, &content).await {
```

Comment:
Can be `.map_err` (or [`.tap_err`](https://docs.rs/tap/latest/tap/trait.TapFallible.html#tymethod.tap_err)).